api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,424 @@
+ * for interacting with the user's SQLite representation.
+ *
+ * @author wspride
+ */
+public class SqlIndexedStorageUtility<T extends Persistable> implements IStorageUtilityIndexed<T>, Iterable<T> {
+
+    private final Class<T> prototype;
+    private final String tableName;
+    private final String userName;
+
+    public SqlIndexedStorageUtility(Class<T> prototype, String userName, String tableName) {
+        this.tableName = tableName;
+        this.userName = userName;
+        this.prototype = prototype;
+        tryCreateTable();
+    }
+
+    Connection getConnection() {
+        try {
+            Class.forName("org.sqlite.JDBC");
+            return DriverManager.getConnection("jdbc:sqlite:" + this.userName + ".db");
+        } catch (Exception e) {
+            System.out.println("couldn't get jdbc sqlite driver");
+        }
+        return null;
+    }
+
+    /* (non-Javadoc)
+    * @see org.javarosa.core.services.storage.IStorageUtility#write(org.javarosa.core.services.storage.Persistable)
+    */
+    public void write(Persistable p) throws StorageFullException {
+        if (p.getID() != -1) {
+            update(p.getID(), p);
+            return;
+        }
+
+        Connection c;
+        try {
+            c = getConnection();
+            int id = SqlHelper.insertToTable(c, tableName, p);
+            c.close();
+
+            c = getConnection();
+            p.setID(id);
+            SqlHelper.updateId(c, tableName, p);
+            c.close();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public T readFromBytes(byte[] mBytes) {
+
+        T t;
+        try {
+            t = prototype.newInstance();
+            ByteArrayInputStream mByteStream = new ByteArrayInputStream(mBytes);
+            t.readExternal(new DataInputStream(mByteStream), PrototypeManager.getDefault());
+            return t;
+        } catch (InstantiationException | IllegalAccessException | DeserializationException | IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /* (non-Javadoc)
+    * @see org.javarosa.core.services.storage.IStorageUtility#read(int)
+    */
+    public T read(int id) {
+        byte[] mBytes = readBytes(id);
+        return readFromBytes(mBytes);
+
+    }
+
+    @Override
+    public Vector getIDsForValue(String fieldName, Object value) {
+        Connection c = null;
+        PreparedStatement preparedStatement = null;
+        try {
+            c = this.getConnection();
+            preparedStatement = SqlHelper.prepareTableSelectStatement(c, this.tableName,
+                    new String[]{fieldName}, new String[]{(String)value}, prototype.newInstance());
+            if (preparedStatement == null) {
+                return null;
+            }
+            ResultSet rs = preparedStatement.executeQuery();
+            if(rs == null){
+                return null;
+            }
```

Comment:
You don't need to perform null checks on the result of `PreparedStatement.executeQuery()`, the docs say that it will never return a null value 
https://docs.oracle.com/javase/7/docs/api/java/sql/Statement.html#executeQuery(java.lang.String)